### PR TITLE
Add SSA and topologically sorted as requirements to ONNX graph

### PR DESCRIPTION
**Pull request by bddppq**

### CODE REVIEW

File: docs/IR.md [hunk 1]
```diff
@@ -38,7 +38,7 @@ Further metadata may be added to a model via its `metadata_props` field (as desc
 
 ### Graphs
 
-Each computation dataflow graph is structured as a list of nodes that form a graph. These nodes MUST be free of cycles. Each node is a call to an operator. Nodes have one or more inputs, one or more outputs, and zero or more attribute-value pairs.
+Each computation dataflow graph is structured as a list of nodes that form a graph. These nodes MUST be free of cycles and the list is in topoligically sorted order. Each node is a call to an operator. Nodes have one or more inputs, one or more outputs, and zero or more attribute-value pairs.
```
Comment (gramalingam) [inline docs/IR.md]: topoligically => topologically

File: onnx/checker.cc [hunk 2]
```diff
@@ -216,25 +216,51 @@ void check_graph(const GraphProto& graph, int ir_version) {
...
-      ex.AppendContext("Bad node spec: " + node.ShortDebugString());
-      throw ex;
-    }
-   }
 
-  std::unordered_set<std::string> input_names{};
+  std::unordered_set<std::string> output_names{};
   for (const auto& value_info : graph.input()) {
-    input_names.insert(value_info.name());
+    output_names.insert(value_info.name());
```
Comment (linkerzhang) [inline onnx/checker.cc]: want to check no duplicate graph input name please?

File: onnx/checker.cc [hunk 3]
```diff
@@ -216,25 +216,51 @@ void check_graph(const GraphProto& graph, int ir_version) {
...
+            "Nodes in a graph must be topologically sorted, however input '",
+            input,
+            "' of node: \n",
+            node.ShortDebugString(),
+            "\n is not output of any previous nodes.");
+      }
+    }
+    // check for SSA form
+    for (const auto& output : node.output()) {
+      if (output_names.count(output)) {
```
Comment (linkerzhang) [inline onnx/checker.cc]: output name could also be empty when it's optional and you may skip the check when it is.

File: onnx/checker.cc (hunk 2 above)
Comment (bddppq) [reply onnx/checker.cc]: Sorry no, because it's not the intention of this PR. Feel free to do it in a separate one.

File: onnx/checker.cc (hunk 3 above)
Comment (bddppq) [reply onnx/checker.cc]: Good catch, fixed.

File: docs/IR.md [hunk 4]
```diff
@@ -38,7 +38,11 @@ Further metadata may be added to a model via its `metadata_props` field (as desc
 
 ### Graphs
 
-Each computation dataflow graph is structured as a list of nodes that form a graph. These nodes MUST be free of cycles. Each node is a call to an operator. Nodes have one or more inputs, one or more outputs, and zero or more attribute-value pairs.
+Each computation dataflow graph is structured as a list of nodes that form a graph. Each node is a call to an operator. Nodes have one or more inputs, one or more outputs, and zero or more attribute-value pairs. Additionally, a valid ONNX graph must fullfil the following requirements:
```
Comment (ezyang) [inline docs/IR.md]: Rereading this... don't constant nodes have no inputs? :)

Comment (bddppq) [reply docs/IR.md]: Yep good catch

**Pull request merged**